File: src/lib/rusto/src/base_when.rs
use crate::types::IbddInAny;
use std::collections::HashMap;

pub struct BaseWhen<I: IbddInAny> {
    pub name: String,
    pub when_cb: Box<dyn Fn(I::Iselection) -> I::Then>,
    pub error: Option<String>,
    pub artifacts: Vec<String>,
    pub status: Option<bool>,
}

impl<I: IbddInAny> BaseWhen<I> {
    pub fn new(name: String, when_cb: Box<dyn Fn(I::Iselection) -> I::Then>) -> Self {
        Self {
            name,
            when_cb,
            error: None,
            artifacts: Vec::new(),
            status: None,
        }
    }
    
    pub fn add_artifact(&mut self, path: String) {
        let normalized_path = path.replace('\\', "/");
        self.artifacts.push(normalized_path);
    }
    
    pub fn to_obj(&self) -> HashMap<String, Box<dyn std::any::Any>> {
        let mut obj = HashMap::new();
        obj.insert("name".to_string(), Box::new(self.name.clone()) as Box<dyn std::any::Any>);
        obj.insert("status".to_string(), Box::new(self.status) as Box<dyn std::any::Any>);
        obj.insert("error".to_string(), Box::new(self.error.clone()) as Box<dyn std::any::Any>);
        obj.insert("artifacts".to_string(), Box::new(self.artifacts.clone()) as Box<dyn std::any::Any>);
        obj
    }
    
    pub async fn test(&mut self, store: I::Istore) -> Result<I::Istore, String> {
        // In a real implementation, this would execute the when callback
        // For now, just return the store unchanged
        Ok(store)
    }
}

File: src/lib/kafe/BaseWhen.java
package kafe;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class BaseWhen {
    public String name;
    public Object whenCb;
    public Exception error;
    public List<String> artifacts;
    public Boolean status;
    
    public BaseWhen(String name, Object whenCb) {
        this.name = name;
        this.whenCb = whenCb;
        this.artifacts = new ArrayList<>();
        this.status = null;
    }
    
    public void addArtifact(String path) {
        if (!(path instanceof String)) {
            throw new RuntimeException(
                "[ARTIFACT ERROR] Expected string, got " + path.getClass().getName() + ": " + path
            );
        }
        String normalizedPath = path.replace('\\', '/');
        artifacts.add(normalizedPath);
    }
    
    public abstract Object andWhen(Object store, Object whenCb, ITTestResourceConfiguration testResourceConfiguration);
    
    public Map<String, Object> toObj() {
        String errorStr = null;
        if (error != null) {
            errorStr = error.getClass().getName() + ": " + error.getMessage();
        }
        Map<String, Object> obj = new HashMap<>();
        obj.put("name", name);
        obj.put("status", status);
        obj.put("error", errorStr);
        obj.put("artifacts", artifacts);
        return obj;
    }
    
    public Object test(Object store, ITTestResourceConfiguration testResourceConfiguration) throws Exception {
        try {
            Object result = andWhen(store, whenCb, testResourceConfiguration);
            status = true;
            return result;
        } catch (Exception e) {
            status = false;
            error = e;
            throw e;
        }
    }
}
